bundles/org.openhab.binding.fritzboxtr064/README.md

```diff
@@ -0,0 +1,103 @@
+
+Tested devices are:
+- AVM Fritzbox 6490 Cable
+
+## Supported Things
+
+Three things are supported:
+
+- `rootDevice`: the internet gateway device itself
+- `subDevice`: a sub-device of a `rootDevice` (e.g. a WAN interface) 
+- `subDeviceLan`: a special type of sub-device that supports MAC-detection
+
+## Discovery
+
+The root device needs to be added manually, after that sub-devices are detected automatically.
+It is highly recommended to configure that way and NOT rely on manual configuration.
+
+## Thing Configuration
+
+All things have a `refresh``parameter.
+It sets the refresh-interval in seconds for each device channel.
+The default value is 60.
+
+### `rootdevice`
+
+The `host` parameter is needed to communicate with the device.
+It can be a hostname or an IP address and defaults to `fritz.box`.
+
+For accessing the device you need to supply credentials.
+If you did only configure password authentication for your device, the `user` parameter needs to be set to `dslf-config`.
+This is also the default value if the parameter is not set.
+The second credential parameter is `password` and is mandatory.
+For security reasons it is highly recommended to set both, username and password.
+
+One or more TAM (telephone answering machines) are supported by most devices.
+By setting the `tamIndices` parameter you can instruct the binding to add channels for these devices to the thing.
+Values start with `0`.
+This is an optional parameter and multiple values are allowed.
+
+Most devices allow to configure call deflections.
+If the `callDeflectionIndices` parameter is set, channels for the status of the pre-configured call deflections are added.
+Values start with `0`, including the number of "Call Blocks" (two configured call-blocks -> first deflection is `2`).
+This is an optional parameter and multiple values are allowed.
+
+Most devices support call lists.
+The binding can analyze these call lists and provide channels for the number of missed calls, inbound calls, outbound calls and rejected (blocked) calls.
+The days for which this analysis takes place can be controlled with the `missedCallDays`, `rejectedCallDays`, `inboundCallDays` and `outboundCallDays`
+This is an optional parameter and multiple values are allowed.
+
+### `subdevice`, `subdeviceLan`
+
+Besides the bridge the thing is attached to, sub-devices have a `uuid` parameter.
+This is the UUID/UDN of the device and a mandatory parameter.
+Since the value can only be determined by examining the SCPD of the root device it is highly recommended to use discovery for adding sub-devices.
+
+For `subdeviceLan` devices (type detected automatically during discovery) a second parameter is available `macOnline`.
+It adds a channel for each MAC (format 11:11:11:11:11:11) that shows the online status of the respective device.
+This is an optional parameter and multiple values are allowed.
+
+## Channels
+
+| channel | item-type | description |
```

Comment:
Surte, but what channels would you like to see as advanced?

Reply:
Actually most of them. I would focus on the channels that are useful for most users in a "standard" setup to be non-advanced, 5-8 should be just fine. Candidates would imho be tamEnable, tamNewMessages, wanConnectionStatus, wifiEnabled.

Reply:
How can I set advanced-attribute on a channel  without using a channel-type? Unfortunately the `ChannelBuilder` knows nothing about `advanced` channels and I don't have channel-types ATM.

Reply:
Err, what do you mean by "not having channel-types"? Every channel should have a proper channel type and the advanced flag belongs to it.

Reply:
Well, I only have dynamic channels. It's possible to create these with all necessary information using the `ChannelBuilder` without using `ChannelTypeUID`s. In fact it's not possible to create channels from `ChannelTypeUID`, you can only create a channel and assign a `ChannelTypeUID` later. The information is duplicated here, you need to specify the accepted item-type in the `ChannelType` and in `ChannelBuilder.create`. I think we should have a `ChanneBuilder.create(ChannelUID, ChannelTypeUID) ` which is far more useful than the `ChannelBuilder.create(ChannelUID, acceptetItemType)` we have now.

I'll implement a `ChannelTypeProvider` to programatically create the needed `ChannelType`s and use those when creating the channels.

Reply:
Hm, interesting indeed, I somehow forgot about the internal details - thanks for pointing it out again.

> you can only create a channel and assign a ChannelTypeUID later

This is not correct. You can (and normally should) use the `withType` method of the `ChannelBuilder` before building the channel.

> I think we should have a ChanneBuilder.create(ChannelUID, ChannelTypeUID) which is far more useful than the ChannelBuilder.create(ChannelUID, acceptedItemType) we have now.

Probably even better a `ChannelBuilder.create(ChannelID, ChannelTypeUID)` (notice the ChannelID instead of ChannelUID) - I agree that the acceptedItemType feels duplicated here as this information is part of the type already. @cweitkamp & @wborn, any input from your side on this?

> I'll implement a ChannelTypeProvider to programmatically create the needed ChannelTypes and use those when creating the channels.

Is there any reason why those channel types cannot simply be statically defined in xml files? The dynamics should only come in when constructing the things and defining their concrete channels.

Reply:
The problem with using channelId instead of channelTypeId is that the Channelbuilder knows nothing about the thing, so it cannot determine the correct thingUID.

To check which channels are available I need additional information besides what is available in the channelType (e.g. the endpoints for the SOAP requests). I would avoid keeping track of channel related information in different files, so I decided to generate the channels on-the-fly. 

Defining the types in an XML would Again require to do that, because every channel would need it‘s own channelType (label is mandatory when defining channelType in an XML). Since the information is already available in my own data, it‘s easier to use a channeltypeprovider.

Reply:
> The problem with using channelId instead of channelTypeId is that the Channelbuilder knows nothing about the thing, so it cannot determine the correct thingUID.

Yeah, you've got a point here.

Wrt the channel types, I do not really agree. I'd assume that we already know all potential types at time of coding and not just at runtime. We will have to document them in some way in the README as well. Having them defined in XML will also allow adding localization, which would otherwise be almost impossible.

Reply:
Yes, we know them at coding time. But we need to keep them in (at least) three files: the .java containing the binding constants, the XML defining the channel-types and the XML defining the possible channels with the information needed for detecting channel availability and providing the endpoints for the requests.

The solution I have know reduces that to one file, which is much easier for extending to other channels and also reduces the risk of introducing coding errors.

Reply:
ok.

Reply:
> Probably even better a `ChannelBuilder.create(ChannelID, ChannelTypeUID)` (notice the ChannelID instead of ChannelUID) - I agree that the acceptedItemType feels duplicated here as this information is part of the type already. @cweitkamp & @wborn, any input from your side on this?

The `acceptedItemType` indeed is superfluous payload. It even is annotated by `@Nullable` in the `ChannelBuilder.create()` [method](https://github.com/openhab/openhab-core/blob/c747f7f9bb062b6b9420ef5b4de44dc17f07ef24/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/binding/builder/ChannelBuilder.java#L57-L66). As @J-N-K already pointed out we at least need a to a Thing and / or a ChannelGroup to create a channel, I guess that is why we only use `ChannelUID` in the ChannelBuilder's create methods. If we want to add more builder methods we need to pass those values too.

Reply:
As follow-up of this discussion I added a shortcut method `Channelbuilder.create(ChannelUID)` in https://github.com/openhab/openhab-core/pull/1572. As a consequence we additionally need a `withAcceptedItemType(String)` method for the builder to cover setting the accepted item types later.